### PR TITLE
[MM-20546] Migrate mfa_section' module and associated tests to TypeScript

### PR DESCRIPTION
**Pull request by hiendinhngoc**
#### Summary
Migrate 'components/user_settings/security/mfa_section' module and associated tests to TypeScript

#### Ticket Link
  Fixes [#13659](https://github.com/mattermost/mattermost-server/issues/13695)

 JIRA: https://mattermost.atlassian.net/browse/MM-20546

### CODE REVIEW

**Comment by hanzei** [comment]
Hi @hiendinhngoc,

Thank you for submitting this PR. Let me know, if you need help resolving the lint failures.

**Comment by hiendinhngoc** [comment]
> Hi @hiendinhngoc,
> 
> Thank you for submitting this PR. Let me know, if you need help resolving the lint failures.

@hanzei could you tell me which file that contains the config rules for lint? I did fix it one time, but it seems doesn't satisfy rule.

**Comment by hanzei** [comment]
@hiendinhngoc You did fix all lint issues :+1: Queuing for developers review

File: components/user_settings/security/mfa_section/mfa_section.tsx [hunk 1]
```diff
@@ -1,46 +1,50 @@
...
+    // Whether or not this server enforces that all users have MFA
+    mfaEnforced: boolean;
 
-        updateSection: PropTypes.func.isRequired,
-        actions: PropTypes.shape({
-            deactivateMfa: PropTypes.func.isRequired,
-        }),
-    };
+    updateSection: (section: string) => void;
+    actions: {deactivateMfa: () => any};
```
Comment (devinbinnie) [inline components/user_settings/security/mfa_section/mfa_section.tsx]: Non-blocking, but this can return `void` since we aren't using the return value.

File: components/user_settings/security/mfa_section/mfa_section.test.tsx [hunk 2]
```diff
@@ -101,21 +101,33 @@ describe('MfaSection', () => {
 
     describe('setupMfa', () => {
         it('should send to setup page', () => {
-            const wrapper = mountWithIntl(<MfaSection {...baseProps}/>);
```
Comment (larkox) [inline components/user_settings/security/mfa_section/mfa_section.test.tsx]: Why did you remove `mountWithIntl`?

File: components/user_settings/security/mfa_section/mfa_section.test.tsx [hunk 3]
```diff
@@ -139,11 +157,17 @@ describe('MfaSection', () => {
...
             const error = {message: 'An error occurred'};
 
-            const wrapper = mountWithIntl(<MfaSection {...baseProps}/>);
+            const wrapper: ShallowWrapper<any, any, MfaSection> = shallow(
+                <MfaSection {...baseProps}/>
+            );
+
+            const mockEvent = {
+                preventDefault: jest.fn()
+            } as unknown as React.MouseEvent<HTMLElement>;
```
Comment (larkox) [inline components/user_settings/security/mfa_section/mfa_section.test.tsx]: I do not like this, but I understand this is the only way to create a MouseEvent object without manually adding the many different properties it must have. Am I right?

File: components/user_settings/security/mfa_section/mfa_section.test.tsx (hunk 2 above)
Comment (hiendinhngoc) [reply components/user_settings/security/mfa_section/mfa_section.test.tsx]: In or der to fix the error `Property 'setupMfa' does not exist on type 'Component<{}, {}, any>'.` then I switch to `ShallowWrapper`

File: components/user_settings/security/mfa_section/mfa_section.test.tsx (hunk 3 above)
Comment (hiendinhngoc) [reply components/user_settings/security/mfa_section/mfa_section.test.tsx]: yes

File: components/user_settings/security/mfa_section/mfa_section.test.tsx (hunk 2 above)
Comment (larkox) [reply components/user_settings/security/mfa_section/mfa_section.test.tsx]: Could you try if making explicit use of the generic solves that problem? What I mean is something like:
`mountWithIntl<MfaSection>(<MfaSection {...baseProps}/>);`

File: components/user_settings/security/mfa_section/mfa_section.tsx (hunk 1 above)
Comment (hiendinhngoc) [reply components/user_settings/security/mfa_section/mfa_section.tsx]: we cannot return void since error is type any.
const {error} = await this.props.actions.deactivateMfa(); line 50

File: components/user_settings/security/mfa_section/mfa_section.test.tsx (hunk 2 above)
Comment (hiendinhngoc) [reply components/user_settings/security/mfa_section/mfa_section.test.tsx]: already tried but still got an error:
```
Type 'MfaSection' does not satisfy the constraint 'ReactElement<any, string | ((props: any) => ReactElement<any, string | ... | (new (props: any) => Component<any, any, any>)> | null) | (new (props: any) => Component<any, any, any>)> | IntlInjectedElement'.
  Type 'MfaSection' is missing the following properties from type 'ReactElement<any, ForwardRefExoticComponent<Pick<WithIntlProps<WrappedComponentProps<any>>, any> & RefAttributes<ComponentType<any>>> & { ...; }>': type, key

(alias) class MfaSection
import MfaSection
```

Comment (larkox) [reply components/user_settings/security/mfa_section/mfa_section.test.tsx]: You are right. The right way to do it would be using this line instead:
`(wrapper.instance() as MfaSection).setupMfa(mockEvent);`

Sorry for the wrong lead before.

File: components/user_settings/security/mfa_section/mfa_section.tsx (hunk 1 above)
Comment (larkox) [reply components/user_settings/security/mfa_section/mfa_section.tsx]: I see the problem. deactivateMfa() should return a Promise with an an object with an object error, which might be undefined, and which has an attribute message. Would you mind checking the changes on the index.js in this PR and follow the same pattern?
You may (or may not) have to migrate the file where deactivateMfa is declared.

https://github.com/mattermost/mattermost-webapp/pull/4837/files

File: components/user_settings/security/mfa_section/mfa_section.tsx [hunk 4]
```diff
@@ -1,46 +1,50 @@
...
+}
 
-    state = {
-        serverError: null,
-    };
+type State = {
+    serverError: string|null;
```
[the remark this answers is not in the document]
Comment (hiendinhngoc) [reply components/user_settings/security/mfa_section/mfa_section.tsx]: I think null is appropriate

File: package-lock.json [hunk 5]
```diff
@@ -5657,7 +5657,7 @@
         },
         "p-cancelable": {
```
Comment (sudheerDev) [inline package-lock.json]: Can you revert the changes to the lock file?

Comment (hiendinhngoc) [reply package-lock.json]: ok I will update it

**Comment by sudheerDev** [comment]
@hiendinhngoc Hey, i still see package-lock.json in the changes. Mind mentioning me when it is reverterd?

**Comment by sudheerDev** [comment]
@lindy65 Up for QA review. You would need to test for mfa in user settings
